### DISCUSSION TITLE
File requirements / parameters

### DISCUSSION THREAD

**Discussion by ElijahElliotP** [semiautomaticgit/SemiAutomaticClassificationPlugin#231] [Q&A] [answered]
Hello,

I'm trying to make use of SCP to clean up or otherwise work with some classifications I am working with in QGIS and and I cannot seem to find all of my layers in the plugin but it's not clear to me why that is.
The layers marked in red are not visible inside the plugin (but I would like them to be).
![image](https://user-images.githubusercontent.com/28412116/180083985-9456d76a-d1c6-4bf8-8ac4-42c7453ac10f.png)
It does not matter which module I choose, it's the same layer set that is available throughout the SCP, and I'm just curious what I should look for that they may have in common.
![image](https://user-images.githubusercontent.com/28412116/180084158-78f670a2-5259-4dec-ba6c-c3c95ca995a5.png)

Any help is appreciated.
Thank you

**Comment by semiautomaticgit** [comment]
Hello @ElijahElliotP ,
the layers listed in SCP menu are single band raster.
Probably the rasters that are not listed are multiband raster.
You can split multiband rasters using this tool https://semiautomaticclassificationmanual.readthedocs.io/en/latest/splitrasterbandsTab.html

**Comment by ElijahElliotP** [comment]
Hi @semiautomaticgit, thank you for taking the time to reply.
Unfortunately this is not the case. Out of an abundance of caution (and because I do appreciate your help), I did attempt to use the tool you recommended but there are no bands listed there. This does not surprise me, because the rasters I'm concerned about are single band thematic maps / classifications. What else might be cause this, do you think? Is there a file size limit, or a bit depth limit?

Cheers,
Eli

**Comment by semiautomaticgit** [comment]
Hi @ElijahElliotP , file size should not be an issue.
The classification list is populated with single band rasters.
What operating system are you using?

**Reply by ElijahElliotP** [reply]
I'm running Windows 10 Version 21H2 build 19044.1889

**Reply by semiautomaticgit** [reply]
Thank you.
Could you please copy the log file following these steps https://semiautomaticclassificationmanual.readthedocs.io/en/latest/faq.html#how-can-i-report-an-error

**Reply by ElijahElliotP** [reply]
Here is the logfile for you!
[SCP_debug_logfile.txt](https://github.com/semiautomaticgit/SemiAutomaticClassificationPlugin/files/9540715/SCP_debug_logfile.txt)

**Reply by semiautomaticgit** [reply]
Thank you for the log file.
There are not errors related to raster list.

I suspect the issue is related to the properties of the rasters that are not listed.
The selection of rasters is based on this function
https://github.com/semiautomaticgit/SemiAutomaticClassificationPlugin/blob/master/core/utils.py#L6522

If a layer is not listed is because it is not a QgsMapLayer.RasterLayer or the raster bandCount() is not 1.

Could you please open the QGIS Python console and run the following command
`qgis.core.QgsProject.instance().mapLayers().values()`

The output could provide information about the layers.

**Reply by ElijahElliotP** [reply]
I copy and pasted the command, run it, and this is the result:
> dict_values([<QgsRasterLayer: '74e-hBULC-2017-MooseSimple-lossProb_gt4_treed_May-Sep' (gdal)>, <QgsRasterLayer: '74e-hBULC-2018-MooseSimple-expectPeriodMean-gt_4' (gdal)>, <QgsRasterLayer: '74e-hBULC-2018-MooseSimple-lossProb_gt4_treed2' (gdal)>, <QgsRasterLayer: '74e_hBULC_Moose_L8L7S2_2expY_2017_May-Sep_20seq_mrg' (gdal)>, <QgsRasterLayer: '74e_hBULC_Moose_L8L7S2_2expY_2017_May-Sep_20seq_mrg_clp' (gdal)>, <QgsRasterLayer: '74e_hBULC_Moose_L8L7S2_2expY_2017_May-Sep_20seq_mrg_clp_dn' (gdal)>, <QgsRasterLayer: '74e_hBULC_Moose_L8L7S2_2expY_2017_May-Sep_20seq_mrg_clp_dn_mskdByLPandTP_bin' (gdal)>, <QgsVectorLayer: '74e_hBULC_Moose_L8L7S2_2expY_2017_May_Sep_20seq_mrg_clp_dn_mskdByLPandTP_bin' (ogr)>, <QgsVectorLayer: '74e_hBULC_Moose_L8L7S2_2expY_2017_May_Sep_20seq_mrg_clp_dn_mskdByLPandTP_bin_fix' (ogr)>, <QgsRasterLayer: '74e_hBULC_Moose_L8L7S2_2expY_2017_May-Sep_20seq_mrg_clp_dn_mskdByLPandTP' (gdal)>, <QgsRasterLayer: '74e_hBULC_Moose_L8L7S2_2expY_2017_May-Sep_25seq_DOY_mrg' (gdal)>, <QgsRasterLayer: '74e_hBULC_Moose_L8L7S2_2expY_2018_May-Sep_25seq_DOY_mrg' (gdal)>, <QgsRasterLayer: '74e_hBULC_Moose_L8L7S2_2expY_2019_May-Sep_25seq_DOY_mrg' (gdal)>, <QgsRasterLayer: '74e_hBULC_Moose_L8L7S2_2expY_2020_May-Sep_25seq_DOY_mrg' (gdal)>, <QgsRasterLayer: '74e_hBULC_Moose_L8L7S2_2expY_2021_May-Sep_25seq_DOY_mrg' (gdal)>, <QgsVectorLayer: 'ForestCuts_Bounds' (ogr)>, <QgsRasterLayer: 'GEE Image' (wms)>, <QgsRasterLayer: 'Google Satellite Hybrid' (wms)>, <QgsVectorLayer: 'MFFP_ForestCuts' (ogr)>, <QgsRasterLayer: 'Simple_Moose_protRegMask_unproductiveMasked_pixels' (gdal)>, <QgsVectorLayer: 'TimeSeriesAnchors' (ogr)>, <QgsVectorLayer: 'VCA_NoDTFields' (ogr)>, <QgsVectorLayer: 'validated_cuts_all — VCA_comments_Union2021-2017_Mar-Oct' (ogr)>, <QgsVectorLayer: 'Validated_Cuts_All_Union_2021-2017_May-Sep_20seq' (ogr)>, <QgsVectorLayer: 'afr_moose_simple' (ogr)>, <QgsVectorLayer: 'ecoforestry_cuts_dates-Cutblock dates July 20-Comments Only' (ogr)>, <QgsVectorLayer: 'unionNarrow2020_2019_2018_2017_sequential' (ogr)>, <QgsVectorLayer: 'validated_cuts_all — v_c_a_NoDTFields' (ogr)>])

**Reply by semiautomaticgit** [reply]
Thank you @ElijahElliotP .
The layers are identified as QgsRasterLayer, therefore the issue should be related to the band count. Maybe the band count for these rasters is not equal to 1.
To check this, you could select one of these layer in the Layers panel, then in QGIS Python console run:
```
layer = iface.activeLayer()
print(str(layer), layer.bandCount())
```

Alternatively, you could try to use QGIS raster calculator to create a copy of one of the not listed rasters and check if this layer is listed in the plugin.

**Comment by ElijahElliotP** [comment]
Really, I should be thanking you for taking so much of your time for this issue.

I ran the code as you've suggested and this is the output:
```layer = iface.activeLayer()
print(str(layer), layer.bandCount())
<QgsRasterLayer: '74e_hBULC_Moose_L8L7S2_2expY_2019_May-Sep_25seq_DOY_mrg' (gdal)> 1
layer = iface.activeLayer()
print(str(layer), layer.bandCount())
<QgsRasterLayer: '74e_hBULC_Moose_L8L7S2_2expY_2020_May-Sep_25seq_DOY_mrg' (gdal)> 1
layer = iface.activeLayer()
print(str(layer), layer.bandCount())
<QgsRasterLayer: '74e_hBULC_Moose_L8L7S2_2expY_2021_May-Sep_25seq_DOY_mrg' (gdal)> 1
layer = iface.activeLayer()
print(str(layer), layer.bandCount())
<QgsRasterLayer: '74e_hBULC_Moose_L8L7S2_2expY_2021_May-Sep_25seq_DOY_mrg' (gdal)> 1
```

**Reply by semiautomaticgit** [reply]
Thank you very much for you cooperation.
It is quite a strange issue.

Could you please load one of the not listed rasters in a new QGIS project and see if it is listed in the SCP menu?

**Reply by ElijahElliotP** [reply]
After making a new session, loading in only a couple problem layers, and checking in SCP, I still do not see them listed.
What else could we check?

**Reply by semiautomaticgit** [reply]
Thank you @ElijahElliotP .
Unfortunately I'm out of options.
Could you share one of the rasters?

**Reply by ElijahElliotP** [reply]
A development: after making a new session and bringing in 5 of the layers that refused to show, I checked SCP again in this new blank session and reported back to you that nothing was listed. At the last moment, I saved the session, and closed it out, figuring that our diagnostic testing might need a very "clean slate" like that session, for other checks.
Today, I went to look at the file size of these rasters in order to send you one, a noticed they were about 25 MB each so I thought it would be a perfect opportunity to use that saved session, clip a raster down, make sure the problem was also affecting the smaller clipped raster and then share it to you. ALL FIVE of the problematic rasters are visible in the SCP, in this minimal project (i.e. no other work was done, not other assets are present). I closed out the minimal session and reopen my actual saved work session (where all this began) and they are visible in the SCP there now as well!!!
![image](https://user-images.githubusercontent.com/28412116/190517753-4bb41eb6-a523-4664-a43e-ccbd860209f5.png)
After months of not working they just appear -I did not create this discussion for some time, since discovering the issue. I only reached out after I could not get the rasters to appear, myself. 
I would love to know what was wrong. I'm glad this is fixed but it's incredibly unsatisfying not knowing what the issue was for so long.

**Reply by semiautomaticgit** [reply]
Well, maybe I overlooked the simplest solution.
Did you press the button ↺ after loading new rasters to refresh layer list?
https://semiautomaticclassificationmanual.readthedocs.io/en/latest/classificationSieveTab.html#classification-sieve-2

**Reply by ElijahElliotP** [reply]
I had previously tried to reload the layer list with that button, yes. 
Nothing, at that time, would provoke those layers to appear.

**Reply by semiautomaticgit** [reply]
OK, thank you for your feedback. 
Then I suspect it is something related with those raster files, but I don't know what.
What is the source of the raster files?

**Reply by ElijahElliotP** [reply]
They're exported from Google Earth Engine.

**Reply by semiautomaticgit** [reply]
Maybe the issue could be related to the format of files exported from Google Earth Engine, but it is quite unclear, especially considering that they suddenly appeared in the list of rasters after you clipped them.

**Comment by ElijahElliotP** [comment, marked as answer]
I also realized the other day that, in the course of our exchange, I have
updated my QGIS installation.
I think at this point I've muddied the waters so much that we can't know
what caused the initial issue, unless I can somehow reproduce it.
I'm sincerely grateful for your time in this matter though.

**Reply by semiautomaticgit** [reply]
You are very welcome @ElijahElliotP , thank you for your cooperation.
